documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,683 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+{
+    public InputReadOnlyList(IReadOnlyList<T> other);
+}
+```
+
+The Silk.NET team wishes to reserve the right to add more constructors to this type as it sees fit.
+
+This exists so that, should the Silk.NET choose to, we can optimize the lookup of elements while ensuring things like indexers are inlined and don't result in a virtual call if our implementation allows us to do so.
+
+**INFORMATIVE TEXT:** For example, for joystick and mouse buttons we could use a fixed-sized bit buffer where each bit represents an individual button: 1 for pressed, 0 for unpressed. But for something like keyboard input where there are a large amount of keys, we can't do that and will likely use `Memory<T>` instead.
+
+# Mouse Input
+
+As discussed earlier, the interface will be very simple.
+
+```cs
+public interface IMouse : IInputDevice
+{
+    MouseState State { get; }
+    ICursorConfiguration Cursor { get; }
+    void SetPosition(Vector2 pos);
+}
+```
+
+`State` is the device state as defined earlier.
+
+`Cursor` contains the cursor configuration. This isn't actually state that the end user can change, and has been made an interface rather than a state struct accordingly.
+
+`SetPosition` allows moving the mouse cursor without the end user physically moving their mouse.
+
+The device state returned by `State` fills out the following structure:
+
+```cs
+public struct MouseState
+{
+    public IMouse Device { get; init; }
+    public MouseButtonState Buttons { get; init; }
+    public ICursorConfiguration Cursor { get; } // forwards to Device.Cursor
+    public Vector2 Position { get; set; } // all sets after the first set forward to Device.SetPosition
+    public Vector2 WheelPosition { get; init; }
+}
+```
+
+For ease-of-use, all APIs on `Device` (other than state) are accessible via the state struct, as indicated in the comments in the API snippet.
+
+`MouseButtonState` is defined as:
+```cs
+public struct MouseButtonState
+{
+    public bool this[MouseButton btn] { get; }
+    public InputReadOnlyList<MouseButton> Down { get; init; }
+    public InputReadOnlyList<MouseButton> Up { get; }
+}
+```
+
+The indexer returns `true` if a particular button is pressed, false otherwise. If the developer wishes to enumerate the button state, they must explicitly enumerate through even the `Down` buttons or `Up` buttons.
+
+`Up` and the indexer will be implemented in terms of `Down`, which is the only property that a backend will need to set.
+
+## Enums
+
+```cs
+public enum MouseButton
+{
+    Unknown,
+    LeftButton,
+    RightButton,
+    MiddleButton,
+    Button4,
+    Button5,
+    Button6,
+    Button7,
+    Button8,
+    Button9,
+    Button10,
+    Button11,
+    Button12,
+    Button13,
+    Button14,
+    Button15,
+    Button16,
+    Button17,
+    Button18,
+    Button19,
+    Button20,
+    Button21,
+    Button22,
+    Button23,
+    Button24,
+    Button25,
+    Button26,
+    Button27,
+    Button28,
+    Button29,
+    Button30,
+    Button31
```

Comment:
Is there a good way to say why this ends at 31? otherwise we should probably specify that we provide these as names (for the respective number), but giving `(MouseButton)212` is valid to retrieve the 212th mouse button

Reply:
Because the USB HID spec (page 61) only allocates up to 4 bits for buttons (3 bits + a device-specific bit which can be used as another button bit).

It is impossible today for there to be more buttons than this. If a use case for more than 31 buttons arises, we should update the API then.